backend/service/auditsink/slack/slack.go

```diff
@@ -104,3 +123,90 @@ func formatText(username string, event *auditv1.RequestEvent) string {
 
 	return messageText
 }
+
+// FormatCustomText parses out the audit event metdata request for the custom slack message text
+func FormatCustomText(message string, event *auditv1.RequestEvent) (string, error) {
+	tmpl, err := template.New("customText").Funcs(funcMap).Parse(message)
+	if err != nil {
+		return "", err
+	}
+
+	data, err := getAuditMetadata(event)
+	if err != nil {
+		return "", err
+	}
+
+	var buf bytes.Buffer
+	if err := tmpl.Execute(&buf, data); err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+// returns the API request/response details saved in an audit event
+func getAuditMetadata(event *auditv1.RequestEvent) (*auditEventMetadata, error) {
+	var requestMetadata map[string]interface{}
+	var responseMetadata map[string]interface{}
+
+	// proto -> json
+	reqJSON, err := protojson.Marshal(event.RequestMetadata.Body)
+	if err != nil {
+		return nil, err
+	}
+	respJSON, err := protojson.Marshal(event.ResponseMetadata.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// json -> map[string]interface{}
+	err = json.Unmarshal(reqJSON, &requestMetadata)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(respJSON, &responseMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	return &auditEventMetadata{
+		Request:  requestMetadata,
+		Response: responseMetadata,
+	}, nil
+}
+
+// for inputs that are type slice/map, returns a formatted slack list OR `N/A` if the slice/map is empty
+func slackList(data interface{}) (string, error) {
+	var text string
+	const listItemFormat = "\n- %v"
+	const mapItemFormat = "\n- %v: %v"
+
+	switch reflect.TypeOf(data).Kind() {
+	case reflect.Slice:
+		s := reflect.ValueOf(data)
+		if s.Len() == 0 {
+			return "`N/A`", nil
+		}
+		// example:
+		// - foo
+		// - bar
+		for i := 0; i < s.Len(); i++ {
+			text += fmt.Sprintf(listItemFormat, s.Index(i).Interface())
+		}
+		return text, nil
+	case reflect.Map:
+		m := reflect.ValueOf(data)
+		if m.Len() == 0 {
+			return "`N/A`", nil
+		}
+		// example:
+		// foo: value
+		// bar: value
+		for _, k := range m.MapKeys() {
+			v := m.MapIndex(k)
+			text += fmt.Sprintf(mapItemFormat, k.Interface(), v.Interface())
+		}
+		return text, nil
+	}
+	return "", errors.New("input type is not Slice or Map")
```

Comment:
i wonder if we should still return a string as the error so the message itself isn't dropped on the floor.

it would be better to get a slack audit event
```
Mike performed Delete on pod123
- Labels: ERR_INPUT_NOT_SLICE_OR_MAP
```

than not seeing the event in slack and getting an error log.

Reply:
yeah good idea, updated